README.md
bayes
=====

**bayes** is a small Python utility to reason about probabilities.
It uses a Bayesian system to extract features, crunch belief updates and
spew likelihoods back. You can use either the high-level functions to
classify instances with supervised learning, or update beliefs manually
with the `Bayes` class.

If you want to simply classify and move files into the most fitting folder, run
this program from the command line passing the root folder path as parameter.
  
High Level
----------

```python
from bayes import classify, classify_file

spams = ["buy viagra", "dear recipient", "meet sexy singles"] # etc
genuines = ["let's meet tomorrow", "remember to buy milk"]
message = "remember the meeting tomorrow"
# Classify as "genuine" because of the words "remember" and "tomorrow".
print classify(message, {'spam': spams, 'genuine': genuines})

# Decides if the person with those measures is male or female.
print classify_normal({'height': 6, 'weight': 130, 'foot size': 8},
                      {'male': [{'height': 6, 'weight': 180, 'foot size': 12},
                                {'height': 5.92, 'weight': 190, 'foot size': 11},
                                {'height': 5.58, 'weight': 170, 'foot size': 12},
                                {'height': 5.92, 'weight': 165, 'foot size': 10}],
                       'female': [{'height': 5, 'weight': 100, 'foot size': 6},
                                  {'height': 5.5, 'weight': 150, 'foot size': 8},
                                  {'height': 5.42, 'weight': 130, 'foot size': 7},
                                  {'height': 5.75, 'weight': 150, 'foot size': 9}]})

# Classifies "unknown_file" as either a Python or Java file, considering
# you have directories with examples of each language.
print classify_file("unknown_file", ["java_files", "python_files"])

# Classifies every file under "folder" as either a Python or Java file,
# considering you have subdirectories with examples of each language.
print classify_folder("folder")
```

Low Level
-------------

```python
from bayes import Bayes

print ' -- Spam Filter --'
# Database with number of sightings of each words in (genuine, spam)
# emails.
words_odds = {'buy': (5, 100), 'viagra': (1, 1000), 'meeting': (15, 2)}
# Emails to be analyzed.
emails = [
          "let's schedule a meeting for tomorrow", # 100% genuine (meeting)
          "buy some viagra", # 100% spam (buy, viagra)
          "buy coffee for the meeting", # buy x meeting, should be genuine
         ]

for email in emails:
    # Start with priors of 90% chance being genuine, 10% spam.
    # Probabilities are normalized automatically.
    b = Bayes([('genuine', 90), ('spam', 10)])
    # Update probabilities, using the words in the emails as events and the
    # database of chances to figure out the change.
    b.update_from_events(email.split(), words_odds)
    # Print the email and if it's likely spam or not.
    print email[:15] + '...', b.most_likely()
    
print ''

print ' -- Spam Filter With Email Corpus -- '

# Email corpus. A hundred spam emails to buy products and with the word
# "meeting" thrown around. Genuine emails are about meetings and buying
# milk.
instances = {'spam': ["buy viagra", "buy cialis"] * 100 + ["meeting love"],
             'genuine': ["meeting tomorrow", "buy milk"] * 100}

# Use str.split to extract features/events/words from the corpus and build
# the model.
model = Bayes.extract_events_odds(instances, str.split)
# Create a new Bayes instance with 10%/90% priors on emails being genuine.
b = Bayes({'spam': .9, 'genuine': .1})
# Update beliefs with features/events/words from an email.
b.update_from_events("buy coffee for meeting".split(), model)
# Print the email and if it's likely spam or not.
print "'buy coffee for meeting'", ':', b

print ''

print ' -- Classic Cancer Test Problem --'
# 1% chance of having cancer.
b = Bayes([('not cancer', 0.99), ('cancer', 0.01)])
# Test positive, 9.6% false positives and 80% true positives
b.update((9.6, 80))
print b
print 'Most likely:', b.most_likely()

print ''

print ' -- Are You Cheating? -- '
results = ['heads', 'heads', 'tails', 'heads', 'heads']
events_odds = {'heads': {'honest': .5, 'cheating': .9},
               'tails': {'honest': .5, 'cheating': .1}}
b = Bayes({'cheating': .5, 'honest': .5})
b.update_from_events(results, events_odds)
print b


def b():
    return Bayes((0.99, 0.01), labels=['not cancer', 'cancer'])

# Random equivalent examples, all achieve the same result.
b() * (9.6, 80)
(b() * (9.6, 80)).opposite().opposite()
b().update({'not cancer': 9.6, 'cancer': 80})
b().update((9.6, 80))
b().update_from_events(['pos'], {'pos': (9.6, 80)})
b().update_from_tests([True], [(9.6, 80)])
Bayes([('not cancer', 0.99), ('cancer', 0.01)]) * (9.6, 80)
Bayes({'not cancer': 0.99, 'cancer': 0.01}) * {'not cancer': 9.6,
                                               'cancer': 80}
```


CHANGES.txt
0.3.1 (2014-5-14)
++++++++++++++++++

- Update metada.


0.3.0 (2014-3-28)
++++++++++++++++++

- Increase stability
- Fix edge cases (e.g. callin .opposite on all zeroes)
- Fix label passing between instances
- Better equality testing for Bayes instances
- Override division operator on Python3
- Add unit tests


0.2.1 (2014-3-16)
++++++++++++++++++

- Allow the user to supply priors in `classify` and `classify_normal`.


0.2.0 (2014-3-16)
++++++++++++++++++

- Python3 support.
- Fix bug in `gaussian_probability` (thanks *cokoder*)


0.1.0 (2013-10-24)
++++++++++++++++++

- Initial release.

bayesian/tests.py
import sys

sys.path.append('../')

import unittest
from bayesian import Bayes, classify, classify_normal

class TestBayes(unittest.TestCase):
    def test_empty_constructor(self):
        with self.assertRaises(ValueError):
            b = Bayes()

    def test_list_constructor(self):
        self.assertEqual(Bayes([]), [])
        self.assertEqual(Bayes(()), [])
        self.assertEqual(Bayes(range(5)), [0, 1, 2, 3, 4])
        self.assertEqual(Bayes({'a': 10, 'b': 50}), [10, 50])
        self.assertEqual(Bayes([10, 10, 20]), [10, 10, 20])
        self.assertEqual(Bayes([('a', 10), ('b', 50)]), [10, 50])
        with self.assertRaises(ValueError):
            b = Bayes([('a', 10), ('b', 50), ('a', 15)])

    def test_get_odds(self):
        b = Bayes({'a': 10, 'b': 50})
        self.assertEqual(b['a'],   10)
        self.assertEqual(b['b'],   50)
        self.assertEqual(b[0], 10)
        self.assertEqual(b[1], 50)

        with self.assertRaises(IndexError):
            b[2]

        with self.assertRaises(ValueError):
            b['c']

    def test_set_odds(self):
        b = Bayes((10, 20, 30))
        b[0] = 50
        b[1] = 40
        b[2] = 30
        self.assertEqual(b, [50, 40, 30])

    def test_opposite(self):
        b = Bayes([0.2, 0.8])
        opposite = b.opposite()
        self.assertEqual(opposite[0] / opposite[1], b[1] / b[0])

        b = Bayes([0.2, 0.4, 0.4])
        opposite = b.opposite()
        self.assertEqual(opposite[0] / opposite[1], b[1] / b[0])
        self.assertEqual(opposite[1] / opposite[2], b[2] / b[1])
        self.assertEqual(opposite[0] / opposite[2], b[2] / b[0])

    def test_normalized(self):
        self.assertEqual(Bayes([]).normalized(), [])
        self.assertEqual(Bayes([2]).normalized(), [1])
        self.assertEqual(Bayes([9, 1]).normalized(), [0.9, 0.1])
        self.assertEqual(Bayes([2, 4, 4]).normalized(), [0.2, 0.4, 0.4])
        self.assertEqual(Bayes([2, 0]).normalized(), [1.0, 0])
        self.assertEqual(Bayes([0, 0]).normalized(), [0.0, 0])

    def test_operators(self):
        b = Bayes([5, 2, 3])
        b *= (2, 2, 1)
        b /= (2, 2, 1)
        self.assertEqual(b, [5, 2, 3])

        self.assertEqual(Bayes([.5, .5]) * (.9, .1), [0.45, 0.05])
        self.assertEqual(Bayes([.5, .5]) / (.9, .1), [5 / 9, 5])

        self.assertEqual(Bayes([.5, .5]) * {'0': 0.9, '1': 0.1}, [0.45, 0.05])
        self.assertEqual(Bayes([.5, .5]) * [('0', 0.9), ('1', 0.1)], [0.45, 0.05])

    def test_equality(self):
        b1 = Bayes([0.5, 0.2, 0.3])
        b2 = Bayes([5, 2, 3])
        b3 = Bayes([5, 2, 5])
        self.assertEqual(b1, b2)
        self.assertNotEqual(b1, b3)
        self.assertNotEqual(b2, b3)

    def test_update(self):
        b = Bayes([1, 2])
        b.update((2, 1))
        self.assertEqual(b, [1, 1])
        b.update((2, 1))
        self.assertEqual(b, [2, 1])
        b.update((2, 0))
        self.assertEqual(b, [1, 0])

    def test_update_from_events(self):
        b = Bayes([1, 1])
        b.update_from_events(['a', 'a', 'a'], {'a': (0.5, 2)})
        self.assertEqual(b, [0.5 ** 3, 2 ** 3])

    def test_update_from_tests(self):
        b = Bayes([1, 1])
        b.update_from_tests([True], [0.9, 0.1])
        self.assertEqual(b, [0.45, 0.05])

        b = Bayes([1, 1])
        b.update_from_tests([True, True, True, False], [0.5, 2])
        self.assertEqual(b, [0.5 ** 2, 2 ** 2])

    def test_most_likely(self):
        b = Bayes({'a': 9, 'b': 1})
        self.assertEqual(b.most_likely(), 'a')
        self.assertEqual(b.most_likely(0), 'a')
        self.assertEqual(b.most_likely(0.89), 'a')
        self.assertIsNone(b.most_likely(0.91))

    def test_is_likely(self):
        b = Bayes({'a': 9, 'b': 1})
        self.assertTrue(b.is_likely('a'))
        self.assertTrue(b.is_likely('a', 0.89))
        self.assertFalse(b.is_likely('a', 0.91))

    def test_conversions(self):
        b = Bayes({'a': 9, 'b': 1, 'c': 0})
        self.assertEqual(b, b.normalized())
        self.assertEqual(b.normalized()['a'], 0.9)
        self.assertEqual(b.opposite().opposite(), b)

    def test_extract_events_odds(self):
        instances = {'spam': ["buy viagra", "buy cialis"] * 100 + ["meeting love"],
                     'genuine': ["meeting tomorrow", "buy milk"] * 100}
        odds = Bayes.extract_events_odds(instances)

        b = Bayes({'spam': 0.9, 'genuine': 0.1})
        b.update_from_events('buy coffee for meeting'.split(), odds)
        self.assertEqual(b.most_likely(0.8), 'genuine')


class TestClassify(unittest.TestCase):
    def test_single(self):
        self.assertEqual(classify('a', {'A': []}), 'A')
        self.assertEqual(classify('a', {'A': ['a']}), 'A')
        self.assertEqual(classify('a', {'A': ['a', 'a']}), 'A')
        self.assertEqual(classify('a', {'A': ['a', 'b']}), 'A')

    def test_basic(self):
        self.assertEqual(classify('a', {'A': ['a'], 'B': ['b']}), 'A')
        self.assertEqual(classify('a a a', {'A': ['a'], 'B': ['b']}), 'A')
        self.assertEqual(classify('a a b', {'A': ['a'], 'B': ['b']}), 'A')
        self.assertEqual(classify('a a b', {'A': ['a', 'a'], 'B': ['b']}), 'A')
        self.assertEqual(classify('a b b', {'A': ['a', 'a'], 'B': ['b']}), 'B')
        self.assertEqual(classify('b b b', {'A': ['a', 'a'], 'B': ['b']}), 'B')

    def test_with_extraction(self):
        self.assertEqual(classify('a', {'A': ['a a a'], 'B': ['b']}), 'A')
        self.assertEqual(classify('a', {'A': ['a', 'a'], 'B': ['b b b']}), 'A')

    def test_sample(self):
        spams = ["buy viagra", "dear recipient", "meet sexy singles"]
        genuines = ["let's meet tomorrow", "remember to buy milk"]
        message = "remember the meeting tomorrow"
        instances = {'spam': spams, 'genuine': genuines}
        self.assertEqual(classify(message, instances), 'genuine')

# Classify File and Classify Folder require too much of a test harness for now.

class TestClassifyNormal(unittest.TestCase):
    def test_single(self):
        self.assertEqual(classify_normal({'a': 100}, {'A': [{'a': 100}]}), 'A')
        self.assertEqual(classify_normal({'a': 100, 'b': 0},
                                         {'A': [{'a': 100, 'b': 0}]}), 'A')

        self.assertEqual(classify_normal({'a': 100, 'b': 0},
                                         {'A': [{'a': 100, 'b': 10}],
                                          'B': [{'a': 50, 'b': 100}]}), None)

    def test_basic(self):
        self.assertEqual(classify_normal({'a': 100, 'b': 0},
                                         {'A': [{'a': 100, 'b': 10},
                                                {'a': 99, 'b': -10}],
                                          'B': [{'a': 50, 'b': 100},
                                                {'a': 70, 'b':90}]}), 'A')

    def test_sample(self):
        instance = {'height': 6, 'weight': 130, 'foot size': 8}
        training = {'male': [{'height': 6, 'weight': 180, 'foot size': 12},
                            {'height': 5.92, 'weight': 190, 'foot size': 11},
                            {'height': 5.58, 'weight': 170, 'foot size': 12},
                            {'height': 5.92, 'weight': 165, 'foot size': 10}],
                   'female': [{'height': 5, 'weight': 100, 'foot size': 6},
                              {'height': 5.5, 'weight': 150, 'foot size': 8},
                              {'height': 5.42, 'weight': 130, 'foot size': 7},
                              {'height': 5.75, 'weight': 150, 'foot size': 9}]}
        self.assertEqual(classify_normal(instance, training), 'female')

if __name__ == '__main__':
    unittest.main()


bayes.py
from collections import defaultdict
import os

def classify(instance, classes_instances, extractor=str.split):
    """
    Using `classes_instances` as supervised learning, classify `instance` into
    one of the example classes. `extractor` is a function to convert instances
    into a list of events/features to be analyzed, which defaults to a simple
    word extraction.
    """
    model = Bayes.extract_events_odds(classes_instances, extractor)
    b = Bayes({class_: 1.0 for class_ in classes_instances})
    b.update_from_events(extractor(instance), model)
    return b.most_likely()

def classify_file(file_, folders, extractor=str.split):
    """
    Classify `file_` into one of `folders`, based on the contents of the files
    already there.  `extractor` is a function to convert file contents
    into a list of events/features to be analyzed, which defaults to a simple
    word extraction.
    """
    classes_instances = defaultdict(list)
    for folder in folders:
        for child in os.listdir(folder):
            child_path = os.path.join(folder, child)
            if os.path.isfile(child_path):
                print child_path
                classes_instances[folder].append(child_path)

    new_extractor = lambda f: extractor(open(f).read())
    return classify(file_, classes_instances, new_extractor)

def classify_folder(folder, extractor=str.split):
    """
    Move every file in `folder` into one of its subfolders, based on the
    contents of the files in those subfolders. `extractor` is a function to
    convert file contents int oa list of events/features to be analyzed, which
    defaults to a simple word extraction.
    """
    subfolders = []
    files = []
    for item in os.listdir(folder):
        if os.path.isdir(item):
            subfolders.append(os.path.join(folder, item))
        else:
            files.append(os.path.join(folder, item))

    for file_ in files:
        classification = classify_file(file_, subfolders, extractor)
        os.rename(file_, os.path.join(folder, classification, file_))


from math import sqrt, pi, exp
def gaussian_distribution(values):
    """
    Given a list of values, returns the (mean, variance) tuple for a
    Gaussian (normal) distribution.
    """
    n = float(len(values))
    mean = sum(values) / n
    variance = sum((value - mean) ** 2 for value in values) / (n - 1)
    return (mean, variance)

def gaussian_probability(sample, distribution):
    """
    Given a sample value and the (mean, variance) distribution,
    return the probability of this sample belonging to the
    distribution.
    """
    mean, variance = distribution
    return (exp((sample - mean) ** 2 / (-2 * variance))
            / sqrt(2 * pi * variance))

def properties_distributions(classes_population):
    """
    Converts classes populations into classes distributions by property.
    {class: [{property: value}]} -> {property: {class: distribution}}
    """
    properties_distributions = defaultdict(dict)
    for class_, population in classes_population.items():
        properties_instances = defaultdict(list)
        for properties in population:
            for property, value in properties.items():
                properties_instances[property].append(value)
        for property, instances in properties_instances.items():
            properties_distributions[property][class_] = gaussian_distribution(instances)
    return properties_distributions

def classify_normal(instance, classes_instances):
    """
    Classify `instance` into one of the classes from `classes_instances`,
    calculating the probabilities from the Gaussian (normal) distribution
    from each classes' instances.

    classes_instances must be of type {class: [{property: value}]}
    """
    distributions = properties_distributions(classes_instances)
    b = Bayes({class_: .5 for class_ in classes_instances})
    for property, value in instance.items():
        classes_distributions = distributions[property]
        b.update({class_: gaussian_probability(value, distribution)
                  for class_, distribution in classes_distributions.items()})
    return b.most_likely()
        

class Bayes(list):
    """
    Class for Bayesian probabilistic evaluation through creation and update of
    beliefs. This is meant for abstract reasoning, not just classification.
    """
    @staticmethod
    def extract_events_odds(classes_instances, event_extractor=str.split):
        """
        Runs function `event_extractor` for every instance in every class in
        `classes_instances` ({class: [instances]}) and returns the odds of each
        event happening for each class.

        The result of this function is meant to be used in a future
        `update_from_events` call.
        """
        small = 0.000001
        events_odds = defaultdict(lambda: defaultdict(lambda: small))
        for class_, instances in classes_instances.items():
            for instance in instances:
                for event in event_extractor(instance):
                    events_odds[event][class_] += 1

        return events_odds

    def __init__(self, value=None, labels=None):
        """
        Creates a new Bayesian belief system.

        `value` can be another Bayes
        object to be copied, an array of odds, an array of (label, odds)
        tuples or a dictionary {label: odds}.

        `labels` is a list of names for the odds in `value`. Labels default to
        the their indexes.
        """
        if value is not None:
            # Convert dictionary.
            if isinstance(value, dict):
                labels = labels or sorted(value.keys())
                value = [value[label] for label in labels]

            # Convert list of tuples.
            elif labels is None and len(value) and isinstance(value[0], tuple):
                labels, value = zip(*value)

        super(Bayes, self).__init__(value)
        self.labels = labels or map(str, range(len(self)))

    def __getitem__(self, i):
        """ Returns the odds at index or label `i`. """
        if isinstance(i, str):
            return self[self.labels.index(i)]
        else:
            return super(Bayes, self).__getitem__(i)

    def __setitem__(self, i, value):
        """ Sets the odds at index or label `i`. """
        if isinstance(i, str):
            self[self.labels.index(i)] = value
        else:
            super(Bayes, self).__setitem__(i, value)

    def _cast(self, other):
        """
        Converts and unknown object into a Bayes object, keeping the same
        labels if possible.
        """
        if isinstance(other, Bayes):
            return other
        else:
            return Bayes(other, self.labels)

    def opposite(self):
        """
        Returns the opposite probabilities.
        Ex: [.7, .3] -> [.3, .7]
        """
        if 0 in self:
            return self._cast(1 if i == 0 else 0 for i in self)
        else:
            return self._cast(1 / i for i in self)

    def normalized(self):
        """
        Converts the list of odds into a list probabilities that sum to 1.
        """
        total = float(sum(self))
        return self._cast(i / total for i in self)

    def __mul__(self, other):
        """
        Creates a new instance with odds from both this and the other instance.
        Ex: [.5, .5] * [.9, .1] -> [.45, .05] (non normalized)
        """
        return self._cast(i * j for i, j in zip(self, self._cast(other)))

    def __div__(self, other):
        """
        Creates a new instance with odds from this instance and the opposite of
        the other.
        Ex: [.5, .5] / [.9, .1] -> [.555, 50.0] (non normalized)
        """
        return self * self._cast(other).opposite()

    def update(self, event):
        """
        Updates all current odds based on the likelihood of odds in event.
        Modifies the instance and returns itself.
        Ex: [.5, .5].update([.9, .1]) becomes [.45, .05] (non normalized)
        """
        self[:] = (self * self._cast(event)).normalized()
        return self

    def update_from_events(self, events, events_odds):
        """
        Perform an update for every event in events, taking the new odds from
        the dictionary events_odds (if available).
        Ex: [.5, .5].update_from_events(['pos'], {'pos': [.9, .1]})
        becomes [.45, .05] (non normalized)
        """
        for event in events:
            if event in events_odds:
                self.update(events_odds[event])
        return self

    def update_from_tests(self, tests_results, odds):
        """
        For every binary test in `tests_results`, updates the current belief
        using the item at the same position in `odds`. If the test is true, use
        the odds as is. If it's false, use it's opposite.
        Ex: [.5, .5].update_from_tests([True], [.9, .1]) becomes [.45, .05]
        (non normalized)
        """
        for result, chance in zip(tests_results, odds):
            if result:
                self.update(chance)
            else:
                self.update(self._cast(chance).opposite())
        return self

    def most_likely(self, cutoff=0.0):
        """
        Returns the label with most probability, or None if its probability is
        under `cutoff`.
        Ex: {a: .4, b: .6}.most_likely() -> b
            {a: .4, b: .6}.most_likely(cutoff=.7) -> None
        """
        normalized = self.normalized()
        max_value = max(normalized)

        if max_value > cutoff:
            return self.labels[normalized.index(max_value)]
        else:
            return None

    def is_likely(self, label, minimum_probability=0.5):
        """
        Returns if `label` has at least probability `minimum_probability`.
        Ex: {a: .4, b: .6}.is_likely(b) -> True
        """
        return self.normalized()[label] > minimum_probability

    def __str__(self):
        items = []
        for label, item in zip(self.labels, self.normalized()):
            items.append('{}: {}%'.format(label, round(item * 100, 2)))
        return 'Bayes({})'.format(', '.join(items))


if __name__ == '__main__':
    import sys
    for folder in sys.argv[1:]:
        print folder
        classify_folder(folder)


bayesian/samples.py
import sys
sys.path.append('../')
from bayesian import Bayes, classify_normal, classify

print(' == High Level Functions == ')

print(' -- Gender Classification -- ')
# Decides if the person with those measures is male or female.
print(classify_normal({'height': 6, 'weight': 130, 'foot size': 8},
                      {'male': [{'height': 6, 'weight': 180, 'foot size': 12},
                                {'height': 5.92, 'weight': 190, 'foot size': 11},
                                {'height': 5.58, 'weight': 170, 'foot size': 12},
                                {'height': 5.92, 'weight': 165, 'foot size': 10}],
                       'female': [{'height': 5, 'weight': 100, 'foot size': 6},
                                  {'height': 5.5, 'weight': 150, 'foot size': 8},
                                  {'height': 5.42, 'weight': 130, 'foot size': 7},
                                  {'height': 5.75, 'weight': 150, 'foot size': 9}]}))

print('')

print(' -- Spam Detection With `Classify` -- ')
spams = ["buy viagra", "dear recipient", "meet sexy singles"] # etc
genuines = ["let's meet tomorrow", "remember to buy milk"]
message = "remember the meeting tomorrow"
# Classify as "genuine" because of the words "remember" and "tomorrow".
print(classify(message, {'spam': spams, 'genuine': genuines}))

# Classifies "unknown_file" as either a Python or Java file, considering
# you have directories with examples of each language.
#print classify_file("unknown_file", ["java_files", "python_files"])

# Classifies every file under "folder" as either a Python or Java file,
# considering you have subdirectories with examples of each language.
#print classify_folder("folder")

print('')

print(' == Low Level Functions == ')

print(' -- Classic Cancer Test Problem --')
# 1% chance of having cancer.
b = Bayes([('not cancer', 0.99), ('cancer', 0.01)])
# Test positive, 9.6% false positives and 80% true positives
b.update((9.6, 80))
print(b)
print('Most likely:', b.most_likely())

print('')

print(' -- Spam Filter With Existing Model --')
# Database with number of sightings of each words in (genuine, spam)
# emails.
words_odds = {'buy': (5, 100), 'viagra': (1, 1000), 'meeting': (15, 2)}
# Emails to be analyzed.
emails = [
          "let's schedule a meeting for tomorrow", # 100% genuine (meeting)
          "buy some viagra", # 100% spam (buy, viagra)
          "buy coffee for the meeting", # buy x meeting, should be genuine
         ]

for email in emails:
    # Start with priors of 90% chance being genuine, 10% spam.
    # Probabilities are normalized automatically.
    b = Bayes([('genuine', 90), ('spam', 10)])
    # Update probabilities, using the words in the emails as events and the
    # database of chances to figure out the change.
    b.update_from_events(email.split(), words_odds)
    # Print the email and if it's likely spam or not.
    print(email[:15] + '...', b.most_likely())

print('')

print(' -- Spam Filter With Email Corpus -- ')

# Email corpus. A hundred spam emails to buy products and with the word
# "meeting" thrown around. Genuine emails are about meetings and buying
# milk.
instances = {'spam': ["buy viagra", "buy cialis"] * 100 + ["meeting love"],
             'genuine': ["meeting tomorrow", "buy milk"] * 100}

# Use str.split to extract features/events/words from the corpus and build
# the model.
model = Bayes.extract_events_odds(instances, str.split)
# Create a new Bayes instance with 10%/90% priors on emails being genuine.
b = Bayes({'spam': .9, 'genuine': .1})
# Update beliefs with features/events/words from an email.
b.update_from_events("buy coffee for meeting".split(), model)
# Print the email and if it's likely spam or not.
print("'buy coffee for meeting'", ':', b)

print('')

print(' -- Are You Cheating? -- ')
results = ['heads', 'heads', 'tails', 'heads', 'heads']
events_odds = {'heads': {'honest': .5, 'cheating': .9},
               'tails': {'honest': .5, 'cheating': .1}}
b = Bayes({'cheating': .5, 'honest': .5})
b.update_from_events(results, events_odds)
print(b)

def b():
    return Bayes((0.99, 0.01), labels=['not cancer', 'cancer'])

# Random equivalent examples
b() * (9.6, 80)
(b() * (9.6, 80)).opposite().opposite()
b().update({'not cancer': 9.6, 'cancer': 80})
b().update((9.6, 80))
b().update_from_events(['pos'], {'pos': (9.6, 80)})
b().update_from_tests([True], [(9.6, 80)])
Bayes([('not cancer', 0.99), ('cancer', 0.01)]) * (9.6, 80)
Bayes({'not cancer': 0.99, 'cancer': 0.01}) * {'not cancer': 9.6,
                                               'cancer': 80}


